File: ui/src/slug.ts
/**
 * Path-slug helpers for the "root + name" connector builder (issue #18).
 * Mirrors cli/slug.py — kept client-side on purpose, the admin API's own
 * path validation is unchanged.
 */

export function slugify(text: string): string {
  return text
    .trim()
    .toLowerCase()
    .replace(/[^a-z0-9]+/g, "-")
    .replace(/^-+|-+$/g, "");
}

/** Best-effort reverse of slugify, for friendly display only. */
export function deslugify(segment: string): string {
  return segment
    .split("-")
    .filter(Boolean)
    .map((w) => w[0].toUpperCase() + w.slice(1))
    .join(" ");
}

export function lastSegment(path: string): string {
  const parts = path.split("/").filter(Boolean);
  return parts[parts.length - 1] ?? "";
}

/** Everything but the last path segment. "/api/Defender/nist" -> "/api/Defender". */
export function parentPath(path: string): string {
  const trimmed = path.replace(/\/+$/, "");
  const idx = trimmed.lastIndexOf("/");
  return trimmed.slice(0, idx) || "/";
}

export interface EndpointLike {
  id: string;
  path: string;
}

export interface EndpointGroup<T extends EndpointLike> {
  parentPath: string;
  root: T;
  children: T[];
}

/**
 * Group endpoints that nest one level under a real, existing endpoint.
 *
 * Grouping requires the shared parent path to be a literal endpoint's own
 * path — two unrelated endpoints like "/api/Defender" and "/api/users"
 * share the ancestor "/api", but "/api" isn't a deliberate root either was
 * created under, so treating that as a group would be a false positive.
 * Requiring the parent to actually exist makes "is this a root?"
 * unambiguous.
 */
export function groupEndpoints<T extends EndpointLike>(
  endpoints: T[]
): { groups: EndpointGroup<T>[]; ungrouped: T[] } {
  const byPath = new Map(endpoints.map((e) => [e.path, e]));
  const childrenByParent = new Map<string, T[]>();
  for (const ep of endpoints) {
    const parent = parentPath(ep.path);
    const list = childrenByParent.get(parent) ?? [];
    list.push(ep);
    childrenByParent.set(parent, list);
  }

  const groups: EndpointGroup<T>[] = [];
  const groupedIds = new Set<string>();

  for (const parent of [...childrenByParent.keys()].sort()) {
    const root = byPath.get(parent);
    if (!root) continue;
    const children = [...childrenByParent.get(parent)!].sort((a, b) => a.path.localeCompare(b.path));
    groups.push({ parentPath: parent, root, children });
    groupedIds.add(root.id);
    children.forEach((c) => groupedIds.add(c.id));
  }

  const ungrouped = endpoints
    .filter((e) => !groupedIds.has(e.id))
    .sort((a, b) => a.path.localeCompare(b.path));

  return { groups, ungrouped };
}

/** Every existing endpoint's own path is a valid root to nest a new child under. */
export function knownRoots<T extends EndpointLike>(endpoints: T[]): string[] {
  return [...new Set(endpoints.map((e) => e.path))].sort();
}

File: cli/slug.py
"""
Path-slug helpers shared by the CLI's `endpoint create --root/--name` and
`endpoint list` grouping. The UI has its own TypeScript equivalent
(ui/src/slug.ts) — see issue #18: this stays a client-side/CLI-side
convenience on purpose, the admin API's own path validation is unchanged.
"""

import re
from dataclasses import dataclass, field
from typing import Optional


def slugify(text: str) -> str:
    """Free text -> a URL-safe path segment.

    "Vulnerability scanning" -> "vulnerability-scanning"
    """
    text = text.strip().lower()
    text = re.sub(r"[^a-z0-9]+", "-", text)
    return text.strip("-")


def deslugify(segment: str) -> str:
    """Best-effort reverse of slugify, for friendly display only.

    "vulnerability-scanning" -> "Vulnerability Scanning"
    """
    words = [w for w in segment.split("-") if w]
    return " ".join(w.capitalize() for w in words)


def last_segment(path: str) -> str:
    parts = [p for p in path.split("/") if p]
    return parts[-1] if parts else ""


def parent_path(path: str) -> str:
    """Everything but the last path segment.

    "/api/Defender/nist" -> "/api/Defender"
    "/api/Defender" -> "/api"
    "/Defender" -> "/"
    """
    trimmed = path.rstrip("/")
    idx = trimmed.rfind("/")
    return trimmed[:idx] or "/"


@dataclass
class EndpointGroup:
    parent_path: str
    root: Optional[dict]
    children: list = field(default_factory=list)


def group_endpoints(endpoints: list[dict]) -> tuple[list[EndpointGroup], list[dict]]:
    """Group endpoints that nest one level under a real, existing endpoint.

    Grouping requires the shared parent path to be a literal endpoint's own
    path — not just any shared path prefix. Two unrelated endpoints like
    "/api/Defender" and "/api/users" share the ancestor "/api", but "/api"
    isn't a deliberate root either of them was created under, it's just an
    incidental shallow prefix — grouping those would produce a false
    grouping out of nowhere. Requiring the parent to actually exist as an
    endpoint makes "is this a root?" unambiguous: it either was created at
    that exact path, or it wasn't.

    Returns (groups, ungrouped_endpoints), both sorted by path for stable,
    deterministic display.
    """
    by_path = {ep["path"]: ep for ep in endpoints}
    children_by_parent: dict[str, list[dict]] = {}
    for ep in endpoints:
        parent = parent_path(ep["path"])
        children_by_parent.setdefault(parent, []).append(ep)

    groups: list[EndpointGroup] = []
    grouped_ids: set[str] = set()

    for parent in sorted(children_by_parent):
        root = by_path.get(parent)
        if root is None:
            continue  # no literal endpoint at the parent path — nothing to root a group on
        children = sorted(children_by_parent[parent], key=lambda e: e["path"])
        groups.append(EndpointGroup(parent_path=parent, root=root, children=children))
        grouped_ids.add(root["id"])
        grouped_ids.update(c["id"] for c in children)

    ungrouped = sorted(
        (ep for ep in endpoints if ep["id"] not in grouped_ids), key=lambda e: e["path"]
    )
    return groups, ungrouped
